apps/argos_aggregation/lib/argos_aggregation/data_sources/thesauri.ex

```diff
@@ -23,60 +25,215 @@ defmodule ArgosAggregation.Thesauri do
     end
   end
 
-  defmodule DataProvider do
+
+
+
+  defmodule DataSourceClient.Http do
     @base_url Application.get_env(:argos_aggregation, :thesauri_url)
 
-    import SweetXml
+    def request_by_date(%Date{} = date) do
+       read_from_url("#{@base_url}/search.rdf?q=&change_note_date_from=#{Date.to_iso8601(date)}")
+    end
+
+    def request_root_level() do
+      "#{@base_url}/hierarchy.rdf?depth=0"
+      |> read_from_url()
+    end
+
+    def request_node_hierarchy(id) do
+      "#{@base_url}/hierarchy/#{id}.rdf?dir=down"
+      |> read_from_url([timeout: 50_000, recv_timeout: 50_000])
+    end
+
+    def request_single_node(id) do
+      "#{@base_url}/#{id}.rdf"
+      |> read_from_url()
+    end
+
+    def read_from_url(url) do
+      url
+      |> HTTPoison.get
+      |> fetch_response
+    end
+
+    def read_from_url(url, options) when is_list(options) do
+      url
+      |> HTTPoison.get([], options)
+      |> fetch_response
+    end
+
+    defp fetch_response({:ok, %{status_code: 200, body: body}}), do: {:ok, body}
+    defp fetch_response({:ok, %HTTPoison.Response{status_code: code, request: req}}) do
+      {:error, "Received unhandled status code #{code} for #{req.url}."}
+    end
+    defp fetch_response({:error, error}), do: {:error, error.reason()}
+
+  end
+
+  defmodule DataSourceClient.Local do
+
+    alias ArgosAggregation.ElasticSearch
+
+    def request_single_node(id) do
+      ElasticSearch.DataProvider.get_doc("concept_#{id}")
+    end
+
+  end
+
+
+  defmodule DataProvider do
     require Logger
 
+    alias ArgosAggregation.ElasticSearch.Indexer
+    alias ArgosAggregation.Thesauri.ConceptParser
+
+    def get_by_date(%Date{} = date) do
+      date
+      |> stream_pages
+      |> Stream.map(fn element ->
+        case element do
+          {:error, _err} -> element
+          _ -> ConceptParser.Search.parse_single_doc(element)
```

Comment:
Mir ist nicht ganz klar, wieso load_next_page() direkt hier als Funktion gibt, dann aber auch dieses Search-Modul existiert, deren Funktionen ziemlich ähnlich klingen bzw. zum Teil gar nicht nach "Suche", wie hier das parsen eines einzelnen Dokuments. 

Kann es sein dass die Benennungen etwas "legacy" aus der altem Ansatz sind? Beim ersten Lesen ist es gerade nicht so ganz klar kommuniziert, wofür das Search Modul alles zuständig ist/wieso seine Funktionen nicht einfach hier im Dataprovider oder dem Parser integriert sind.

Reply:
Der Unterschied liegt in der Gestaltung der rdf Dokumente die geparst werden müssen. Die Daten die bei einer Suche zurückkommen sind anders Strukturiert als bei einer Hierarchie. Um die ID eines Concepts in einer Update Suche zu erfahren muss `xpath(~x"//sdc:link/@rdf:resource")` abgefragt werden, in einer Hierarchy `xpath(~x"//@rdf:about"s)`Dadurch habe ich am Ende sehr Domänen spezifische Aufrufe, die ich der Ordnung halber in eigene Sub-Module aufgeteilt habe.